### PR TITLE
WIP feature(amd): Allow AMD modules to be decorated by plugins

### LINKED ISSUE #8082: Allow plugins to decorate AMD modules
``` js
require.config({
  paths: {
    "_alias/foo": "foo.js"
  },
  map: {
    'PLUGIN2/decorator/foo': {
      "foo": "PLUGIN1/decorator/foo"
    },
    'PLUGIN1/decorator/foo': {
      "foo": "_alias/foo"
    },
    '*': {
      "foo": "PLUGIN2/decorator/foo"
    },
  }
});
```

Plugins decorate:

``` js
// PLUGIN1/decorator/foo.js
define(function (require) {
  // due to the map, this actually receives _alias/foo (located at foo.js)
  var foo = require('foo');
  // make changes
  return foo;
});

// PLUGIN2/decorator/foo.js
define(function (require) {
  // due to the map, this actually receives PLUGIN1/decorator/foo
  var foo = require('foo');
  // make changes
  return foo;
});
```

``` js
require(['foo'], function (foo) {
  // due to the map, this actually receives PLUGIN2/decorator/foo
});
```

### PR DESCRIPTION
**Pull request by mrclay**
Includes [docs](https://github.com/mrclay/Elgg-leaf/blob/08f75b8b2291f6cc8e47c81843bf519c37a6a5af/docs/guides/javascript.rst#altering-the-return-values-of-other-modules).

To see it working I extended elgg/ckeditor/config in the blog plugin. The JS error CKEditor creates is not caused by this change.

Fixes #8082

### CODE REVIEW

**Comment by beck24** [comment]
Awesome, can't wait to start debugging js that's gone through multiple levels of modification ;)
LGTM

**Comment by mrclay** [comment]
The good thing is (for now) you can see the scripts requested and the file names make it somewhat apparent what's going on.

**Comment by hypeJunction** [comment]
Me likes. Thanks @mrclay!

File: docs/guides/javascript.rst [hunk 1]
```diff
@@ -43,8 +43,23 @@ A basic module could look like this:
...
+modules.
+
+To register a module named ``my/module``, create the view ``views/default/js/my/module.js.php``
+(note the extension ``.js.php``). You must also manually register your view as an external
+resource:
+
+.. code-block:: php
+
+    // note the view name includes the .js extension, but not .php
+    elgg_register_external_view('js/my/module.js', true);
```
Comment (ewinslow) [inline docs/guides/javascript.rst]: What is the true?

**Comment by ewinslow** [comment]
This is pretty good. applyDecorations requirement is a little strange -- why can't this just happen in getConfig?

**Comment by ewinslow** [comment]
Still feel uneasy with it, but can't put my finger on exactly why right now.

**Comment by ewinslow** [comment]
I guess it's that I see amd modules in the same vein as java packages or php classes. They are supposed to be static and not modifiable/decoratable at run time.

This blends the worlds of di and code loading, and people are already too confused about how that is supposed to work. I'm leaning on my experience with angular here.

**Comment by mrclay** [comment]
Our requirements seem to be that PHP doesn't output/alter JS, and plugins can do little but add/remove AMD modules. I'm not sure these are realistic as JS dev seems to involve a lot of manually wiring modules together. The mix of AMD and hooks I find yucky even though we have kind of a solution.

**Comment by mrclay** [comment]
Closed because it was on 1.x, but excitement seems low.

**Pull request closed without merging**